opengovsg/plumber
# PLU-125: [EXCEL-3] Implement system-added connections for M365 servers

## Problem
We need to constrain what M365 servers our users can connect to. In general, we want to stick to these guidelines:
1. Local development and Govtech staging servers should only be accessible by default during er... local development.
2. Only users from agencies with accounts on M365 should be able to access the SG govt M365 server.

## Solution
Make M365 a system-added connection app; implement its `getSystemAddedConnections` like so:
* For each possible M365 tenant, check against a LD flag named in this naming convention `backend_m365_tenant_{tenantKey}`.
* For convenience, also expose the local development and govtech staging M365 tenants if the "is dev" flag is set.
* If the connection object was not created for the user already, create it on the fly.

Some notes:
* To populate the LD flag, we will manually check if an agency has accounts on SG govt m365 and add their email domains to LD.
* Creating a connection on-the-fly is not conventional (since we are doing DB writes in a GraphQL query), but I think it should be ok to do:
  * it simplifies logic _a lot_,
  * It's (mostly) idempotent. _Mostly_ because i don't deal with concurrency in this PR yet - we don't expect users to perform connection querying concurrently.
* To prevent confusion, we still allow users to see old M365 connections, even if they have become ineligible for it (e.g. we blocked them on LD). They will instead get an error message if they are ineligible (implemented in a later PR). 

**Connections list with open.gov.sg user**
_Note: only local dev + govtech staging, since open.gov.sg domain does not have access to SG Govt m365._
<img width="500" src="https://github.com/opengovsg/plumber/assets/135598754/43a5c745-bc64-426f-a247-bfa58568bb97" />

**Connections list with govtech user**
_Screenshot from local dev so you also see the staging and local dev entries_
<img width="500" src="https://github.com/opengovsg/plumber/assets/135598754/6b2d0a34-4449-4ad1-aeb9-93f72e172233" />


## Tests
- Check that local development and govtech staging servers always show up if the "is development" flag is true.
- Check that SG govt m365 server does not show up if launch darkly flag evaluates to false
- Check that SG govt m365 server show ups if launch darkly flag evaluates to true.

**Unit tests to be added in a separate PR after team is OK with approach**

## Deploy Notes
**New environment variables**:
- `LAUNCH_DARKLY_SDK_KEY`: SDK key for server-side launch darkly

**New dependencies**:
- `@launchdarkly/node-server-sdk`: for server-side launch darkly.
